#179/create register page

fizcamposancos opened the pull request:
closes #179

With the design issues we are discussing in the login page #177

Damrod commented on webapp/src/pages/RegisterPage/RegisterPage.tsx:
```diff
@@ -0,0 +1,161 @@
+import './RegisterPage.css';
+
+export default function RegisterPage() {
+  const [username, setUsername] = useState('');
```
couldn't we use something like this:
```
const [userRegistrationDto, setUserRegistrationDto] = useState<RegisterUserDto>({username: '', fullName: '', email: '', password: '', confirmationPassword: ''});
```
instead of these some 5 useState, since we already have the type anyway?

fizcamposancos replied:
I've finally used `useReducer` instead. [https://reactjs.org/docs/hooks-reference.html#usereducer](https://reactjs.org/docs/hooks-reference.html#usereducer)

Damrod replied:
To be honest I think this is not the _by the book_ ideal use case for useReducer (a reducer becomes a real help, and almost a necessity, when we have to handle complex state logic. Here we don't have complex state logic: we're only updating the dto).

My idea was something like what you can find here `webapp/src/shared/hooks/UseDrag.tsx` L21.

But it works, so I'll just approve it.

Damrod commented on webapp/src/pages/RegisterPage/RegisterPage.css:
```diff
@@ -0,0 +1,72 @@
+  .register .landing-animation {
+    display: inline;
+  }
+}
```
I have a problem, again, with the height of the page. I work on a small screen, and have to do quite some scrolling even in a page with not so much content as this one, but maybe there's no clean way around my scrolling...
Where even is the figma for this particular page? Did you wing it?

fizcamposancos replied:
Yes, the ticket is open, once I fix the other one I will fix this as well. BTW I couldn't do what you suggested, `scroll: clip `, what did you mean?

Damrod replied:
I'm not sure. It was pseudo code, not really code. I will try to find a solution for this tomorrow.

Damrod commented on webapp/src/pages/RegisterPage/RegisterPage.css:
```diff
@@ -0,0 +1,72 @@
+  color: var(--color-light);
+  display: grid;
+  grid-template-columns: repeat(var(--column-number), 1fr);
+  grid-template-rows: auto 1fr;
```
have you noticed that we have been building all pages with a column gap that you are nor having not here not in the login page? V.g:
![Screenshot from 2022-09-15 23-36-38](https://user-images.githubusercontent.com/2774420/190504159-85f15473-eb0d-4e19-bde5-7b8ef6048acf.png)
![Screenshot from 2022-09-15 23-35-38](https://user-images.githubusercontent.com/2774420/190504162-b43cd604-ab92-449c-9909-d3830de5605b.png)

fizcamposancos replied:
No, sorry, I'll fix it

Damrod replied:
I'm sorry, I made a mess with your pr. I wanted to delete the screenshots I posted, and instead I removed the css file. Should be fixed now (I forced pushed your commit from before I created my stupid commit)

fizcamposancos replied:
Don't worry. But I don't know why it's happening. That gap is in the root.css, isn't it? Shouldn't it be there by deafult?

fizcamposancos replied:
Done

Damrod commented on webapp/src/pages/RegisterPage/RegisterPage.tsx:
```diff
@@ -0,0 +1,161 @@
+  }
+  const handleOnSubmit = (e: React.FormEvent<HTMLFormElement>): void => {
+    e.preventDefault();
+    register();
```
```
Promise returned from register is ignored 
```
couldn't we attach a `.catch((e)=>console.error(e));` to `register()` ?

fizcamposancos replied:
Yes, I promise I thought I did it, maybe I didn't push. I add it

Damrod commented:
Now below the title there is a margin + a gap which upon being visualized look exactly the same. Either make the gap between the rows larger and remove the margin, or make the gap between rows none, and let the margin stay the way it is.

Damrod commented:
what about we attempt to use the grid template columns to set the width of the different components that go from some column to a different one? Cf. [here](https://github.com/42AGV/ft_transcendence/pull/188#issuecomment-1249930590)

fizcamposancos commented:
> Now below the title there is a margin + a gap which upon being visualized look exactly the same. Either make the gap between the rows larger and remove the margin, or make the gap between rows none, and let the margin stay the way it is.

I removed the margin and modified the gap. Please, take a look at it.

vdedios commented:
Nice! but I think we should review feedback messages. Right now it seems we have the same "invalid email or passwords" message for all these situations:

- If passwords are different
- If same specific mandatory field is missing

fizcamposancos commented:
> Nice! but I think we should review feedback messages. Right now it seems that we have the same "invalid email or passwords" message for all these situations:
> 
>     * If passwords are different
> 
>     * If same specific mandatory field is missing

I did some extra input validation, but manually. If you prefer to use something like useForm, please let me know and I will change it. Right now I think it cover most of the possible situations

gbudau commented on webapp/src/pages/RegisterPage/RegisterPage.tsx:
```diff
@@ -0,0 +1,194 @@
+        navigate(LOGIN_OPTIONS_URL);
+      }, 2000);
+    } catch (error) {
+      if (userRegisterUserDto.username === '') {
+        setStatus({
+          type: 'error',
+          message: 'Username can not be empty',
+        });
+      } else if (userRegisterUserDto.email === '') {
+        setStatus({
+          type: 'error',
+          message: 'Email can not be empty',
+        });
+      } else if (userRegisterUserDto.fullName === '') {
+        setStatus({
+          type: 'error',
+          message: 'Full Name can not be empty',
+        });
+      } else if (userRegisterUserDto.password === '') {
+        setStatus({
+          type: 'error',
+          message: 'Password can not be empty',
+        });
+      } else if (
+        userRegisterUserDto.password !==
+        userRegisterUserDto.confirmationPassword
+      ) {
+        setStatus({
+          type: 'error',
+          message: 'Passwords are different',
+        });
```
suggestion: We could extract this part into a function and validate the form values on the client side before making the API call.

Something like this:
```tsx
async function register() {
  if (!hasValidFormValues()) {
    return;
  }
  try {
    ...
  } catch (error) {
    ...
  }
}
```

gbudau reviewed (approved):
Nicely done, thank you! LGTM

Damrod commented on webapp/src/pages/RegisterPage/RegisterPage.tsx:
```diff
@@ -0,0 +1,194 @@
+      ...updateUserRegisterUserDto,
+    }),
+    initialuserRegisterUserDto,
+  );
```
I already mentioned this [here](https://github.com/42AGV/ft_transcendence/pull/180#discussion_r973507804)

Damrod replied:
An [article](https://blog.thoughtspile.tech/2021/10/11/usestate-object-vs-multiple/) on the topic of using multiple state variables vs single state object.

> 
useState calls are more expensive than object allocations, so for 3+ items useState(object) wins

gbudau replied:
@Damrod, thanks! It was a good read.

gbudau commented on webapp/src/pages/RegisterPage/RegisterPage.tsx:
```diff
@@ -0,0 +1,194 @@
+  }
+  const navigate = useNavigate();
+  async function register() {
+    if (hasValidFormValues()) {
```
```suggestion
    if (!hasValidFormValues()) {
```

fizcamposancos replied:
Sorry, thanks!
